mongodb/leafygreen-ui
# [INTEGRATION] Shared config

Creates the following packages
• `@lg-tools/build`: Configuration for building packages. Also includes `TSDoc` build script
• `@lg-tools/create`: Creates a new `@leafygreen-ui` package (improvements to come later)
• `@lg-tools/test`: Configuration for jest tests
• `@lg-tools/lint`: Configuration for esLint & Prettier
• `@lg-tools/storybook`: Configuration for Storybook
• `@lg-tools/storybook-decorators`: Decorators for Storybook
• `@lg-tools/slackbot`: Release announcements slackbot
• `@lg-tools/update`: Updates all `@leafygreen-ui` packages (improvements to come later)
• `@lg-tools/validate`: Build & Dependency validation scripts

Additionally, creates a single package (`@lg-tools/cli`) that builds a Command Line interface. 
This package is intended to be shared across other LeafyGreen repositories, as well as used in projects that use LeafyGreen UI components.

Current supported subcommands include: 

```
lg <subcommand>
  create [options] <name>                 Creates a new package with the provided name
  install [options] [packages...]         Installs LeafyGreen packages to the current app
  link [options] <destination>            Link local LeafyGreen packages to a destination app.
  unlink [options] <destination>          Unlink local LeafyGreen packages from a destination app.
  lint [options]                          Checks code and package.json formatting
  test [options]                          Tests leafygreen-ui packages with unified config.
  update|upgrade [options] [packages...]  Updates installed LeafyGreen packages based on the specified range
  validate [options]                      Validates build integrity, and package.json dependencies
  build-package                           Builds a package
  build-ts                                Builds a package's TypeScript definitions
  build-tsdoc                             Builds a package's TSDoc file
  help [command]                          display help for command
``` 

### Architecture

Inside `tools/cli` we use Commander to create the `lg` command, and all subcommands. 

`cli` depends on several other `@lg-tools` packages. When a subcommand is called, we fire an action function defined in one of the other `@lg-tools/` dependencies.

We define a `bin` property in `cli/package.json` which declares what script to run when we call `> lg` from the terminal


## How to test
Ensure each of the following commands runs without errors (most of these commands are run by CI)
- `yarn build`
- `yarn build:docs`
- `yarn start`
- `yarn lint`
- `yarn run link <path>`
- `yarn test` & `yarn test:ssr`
- `yarn run unlink <path>`
- `yarn validate`
